### ISSUE TITLE
Changing port in development can cause `MongoError`

### ISSUE THREAD

**Issue by klaussner** [meteor/meteor#10381]
The Meteor Tool parses the standard output of the MongoDB process at startup to determine when the development database is ready to accept reads and writes. However, if the app/database port is changed between two `meteor` runs, [`initiateReplSetAndWaitForReady`](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L585) returns too early. If the app accesses the database immediately after it started, this will cause a `MongoError: not master and slaveOk=false`.

Reproduction: https://github.com/klaussner/meteor-issues/tree/10381

I looked into the functions that are responsible for starting MongoDB: First, Meteor spawns the database process in [`launchOneMongoAndWaitForReadyForInitiate`](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L418) and waits until [certain messages](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L535-L549) appear on the standard output. After that, [`initiateReplSetAndWaitForReady`](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L585) initiates or reconfigures the replica set and waits until the database node has "PRIMARY" status. I tested different scenarios (new database, existing database, and existing database with a different port) and recorded the execution order and standard output messages:

**New database (after `meteor reset`):**

```
1. launchOneMongoAndWaitForReadyForInitiate:
    a. MongoDB: [initandlisten] Did not find local replica set configuration document at startup
    b. MongoDB: [initandlisten] waiting for connections on port 3001
    c. "ready to talk"

2. initiateReplSetAndWaitForReady:
    a. initiate/reconfigure replica set
    b. wait until first member is primary

3. Application code: ✅
```

**Existing database:**

```
1. launchOneMongoAndWaitForReadyForInitiate:
    a. MongoDB: [initandlisten] waiting for connections on port 3001
    b. MongoDB: [rsSync-0] transition to primary complete; database writes are now permitted
    c. "ready to talk"

2. initiateReplSetAndWaitForReady:
    a. initiate/reconfigure replica set
    b. wait until first member is primary

3. Application code: ✅
```

**Existing database, different port:**

```
1. launchOneMongoAndWaitForReadyForInitiate:
    a. MongoDB: [initandlisten] waiting for connections on port 4001
    b. MongoDB: [replexec-0] Locally stored replica set configuration does not have a valid entry for the current node
    c. "ready to talk"

2. initiateReplSetAndWaitForReady:
    a. initiate/reconfigure replica set
    b. wait until first member is primary

3. Application code:
    a. MongoError ⚡💢
    b. MongoDB: [rsSync-0] transition to primary complete; database writes are now permitted
```

In the last scenario, the application code is executed too early because `initiateReplSetAndWaitForReady` returns before the `transition to primary complete` message appears on the standard output.

**Comment by benjamn** [comment]
@klaussner Awesome job investigating this! Did your findings point to any possible solution? I've definitely seen the `MongoError: not master and slaveOk=false` many times, and would love to make it stop.

**Comment by klaussner** [comment]
The root of the problem seems to be that `replSetGetStatus` returns `status.members[0].stateStr == "PRIMARY"` ([here](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L664-L665)) before the transition is complete. This has been addressed before in #9119 but the fix doesn't work in this case because the node already has primary status when `initiateReplSetAndWaitForReady` is called, so [`utils.sleepMs(status.heartbeatIntervalMillis)`](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L698) is never executed.

A more reliable way to check if the database is readable/writable is the [`isMaster` command](https://docs.mongodb.com/manual/reference/command/isMaster/). From the docs:

> `isMaster.ismaster`
>
> A boolean value that reports when this node is writable. If `true`, then this instance is a primary in a replica set, or a `mongos` instance, or a standalone `mongod`. This field will be `false` if the instance is a secondary member of a replica set or if the member is an arbiter of a replica set.

I quickly tested this by replacing the [`while` loop in `initiateReplSetAndWaitForReady`](https://github.com/meteor/meteor/blob/0fcc7ddd46d0ef8a278376ba64538210486ab646/tools/runners/run-mongo.js#L663-L716) with


```js
while (!stopped) {
  const { ismaster } = yieldingMethod(db.admin(), "command", {
    isMaster: 1
  });

  if (ismaster) {
    break;
  }

  utils.sleepMs(200);
}
```
